applications/pelion_client/src/modules/net_ot.c

```diff
@@ -181,26 +182,54 @@ static void connect_ot(void)
 	LOG_INF("OT connection requested");
 }
 
-static bool event_handler(const struct event_header *eh)
+static void net_ot_init(bool factory_reset)
 {
-	if (is_module_state_event(eh)) {
-		struct module_state_event *event = cast_module_state_event(eh);
+	static bool initialized;
+
+	__ASSERT_NO_MSG(!initialized);
+	initialized = true;
+
+	if (factory_reset) {
+		struct openthread_context *ot_context = openthread_get_default_context();
+		otError err;
+
+		LOG_WRN("Storage reset requested");
+		openthread_api_mutex_lock(ot_context);
+		err = otInstanceErasePersistentInfo(ot_context->instance);
+		openthread_api_mutex_unlock(ot_context);
+		/* It can fail only if called with OpenThread stack enabled.
+		 * This event should not appear after the OpenThread is started.
+		 * If it does - there is some huge coding error.
+		 */
+		__ASSERT_NO_MSG(err == OT_ERROR_NONE);
+	}
 
-		if (check_state(event, MODULE_ID(main), MODULE_STATE_READY)) {
-			static bool initialized;
+	k_delayed_work_init(&connecting_work, connecting_work_handler);
 
-			__ASSERT_NO_MSG(!initialized);
-			initialized = true;
+	set_net_state_event(NET_STATE_DISCONNECTED);
 
-			k_delayed_work_init(&connecting_work, connecting_work_handler);
+	connect_ot();
+	module_set_state(MODULE_STATE_READY);
+	k_delayed_work_submit(&connecting_work, K_NO_WAIT);
+}
 
-			set_net_state_event(NET_STATE_DISCONNECTED);
+static bool event_handler(const struct event_header *eh)
+{
+	if (!IS_ENABLED(CONFIG_PELION_CLIENT_INIT_FACTORY_RESET_EVENTS) &&
+	    is_module_state_event(eh)) {
+		struct module_state_event *event = cast_module_state_event(eh);
 
-			connect_ot();
-			module_set_state(MODULE_STATE_READY);
-			k_delayed_work_submit(&connecting_work, K_NO_WAIT);
+		if (check_state(event, MODULE_ID(main), MODULE_STATE_READY)) {
+			net_ot_init(false);
```

Comment:
can't we simply send the is_init_factory_reset_event with false if it was disabled?

Reply:
Can we even send an event that is disabled? It seems to be not consistent with the rest of the events.

Reply:
I meant to compile the module in. Instead of waiting for reset source to confirm itself you could emit the event rightaway.

There is also one more thing. The name of the event suggest that it should be sent when reset is requested. In other case there should be no event. I am wondering if we could somehow use this...

Reply:
This way you would have a static dependency chain between modules and you would have a dynamic reaction on reset event.
Reset event would not be needed to initialize the module.

Reply:
e.g. something like below

in modules/factory_reset.c
```
...wiating_done(){
   if (button_held) {
      EVENT_SUBMIT(new_reset_request_event());
   }
   module_set_state(READY);
}
... init()
{
   if (IS_ENABLED(CONFIG_...BUTTON_RESET_ON_INIT_ENABLED)) {
       start_waiting();
   } else {
      module_set_state(READY)
   } 
}
```

in any other module waiting for potential reset
```
if (is_module_state()) {
   if (check_state(event, MODULE_ID(factory_reset), MODULE_STATE_READY)) {
      initme();
      if (reset_requested) {
         resetme();
      } 
   }
}
if (is_reset_request()) {
   reset_requested = true;
}
```

Reply:
Rewritten